### PR TITLE
Globalstorage

### PR DESCRIPTION
**Pull request by MrChico**
Implements the ability to have symbolic storage for multiple contracts by creating fresharrays in Fetch.hs and using the new `sArrayList` technology from sbv

### CODE REVIEW

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 1]
```diff
@@ -725,6 +732,8 @@ symvmFromCommand cmd = do
 
     _ -> error "incompatible options: calldata and abi"
 
+  let model = fromMaybe SymbolicS (storageModel cmd)
```
Comment (xwvvvvwx) [inline src/hevm/hevm-cli/hevm-cli.hs:735]: Is this used anywhere?

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 2]
```diff
@@ -791,6 +800,7 @@ symvmFromCommand cmd = do
       , EVM.vmoptSchedule      = FeeSchedule.istanbul
       , EVM.vmoptChainId       = word chainid 1
       , EVM.vmoptCreate        = create cmd
+      , EVM.vmoptStorageModel  = ConcreteS
```
Comment (xwvvvvwx) [inline src/hevm/hevm-cli/hevm-cli.hs:803]: Shoudl this be `model` or `fromMaybe SymbolicS (storageModel cmd)`?

File: src/hevm/src/EVM.hs [hunk 3]
```diff
@@ -1911,8 +1938,12 @@ create self this xGas xValue xs newAddr initCode = do
         touchAccount self
         touchAccount newAddr
         let
+          store = case view (env . storageModel) vm0 of
+            ConcreteS -> Concrete mempty
+            SymbolicS -> Symbolic $ sListArray (Just 0) []
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM.hs:1943]:
```suggestion
            SymbolicS -> Symbolic $ sListArray Nothing []
```

File: src/hevm/hevm-cli/hevm-cli.hs (hunk 2 above)
Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:803]: yea, this should be `model`. Nice catch

Comment (MrChico) [inline src/hevm/hevm-cli/hevm-cli.hs:803]:
```suggestion
      , EVM.vmoptStorageModel  = model
```

File: src/hevm/src/EVM.hs (hunk 3 above)
Comment (MrChico) [reply src/hevm/src/EVM.hs:1943]: This is creating a new contract, so it's storage is assumed to be empty. We keep storage fully symbolic only if the account has already been initialized previous to this transaction; i.e. when it's loaded from rpc, is the initial contract, or loaded from state (currently not yet supported)

Comment (xwvvvvwx) [reply src/hevm/src/EVM.hs:1943]: ah had missed that this was in `create` :+1:

**Pull request merged**